src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp

```diff
@@ -0,0 +1,223 @@
+#include <Storages/MergeTree/MergeTreeBackgroundExecutor.h>
+
+#include <Common/setThreadName.h>
+#include <Storages/MergeTree/BackgroundJobsAssignee.h>
+
+
+namespace DB
+{
+
+
+String MergeTreeBackgroundExecutor::toString(Type type)
+{
+    switch (type)
+    {
+        case Type::MERGE_MUTATE:
+            return "MergeMutate";
+        case Type::FETCH:
+            return "Fetch";
+        case Type::MOVE:
+            return "Move";
+    }
+    __builtin_unreachable();
+}
+
+
+void MergeTreeBackgroundExecutor::updateConfiguration()
+{
+    auto new_threads_count = threads_count_getter();
+    auto new_max_tasks_count = max_task_count_getter();
+
+    try
+    {
+        pending.set_capacity(new_max_tasks_count);
+        active.set_capacity(new_max_tasks_count);
+
+        pool.setMaxFreeThreads(0);
+        pool.setMaxThreads(new_threads_count);
+        pool.setQueueSize(new_max_tasks_count);
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+
+    threads_count = new_threads_count;
+    max_tasks_count = new_max_tasks_count;
+}
+
+
+void MergeTreeBackgroundExecutor::wait()
+{
+    {
+        std::lock_guard lock(mutex);
+        shutdown = true;
+        has_tasks.notify_all();
+    }
+
+    if (scheduler.joinable())
+        scheduler.join();
+
+    /// ThreadPool will be finalized in destructor.
```

Comment:
I would expect that `wait()` finishes threads from pool, otherwise there may be task running after it.

Reply:
`finalize` is more convenient name for this method...